### PR TITLE
WFLY-10832 Determine the primary provider of singleton service via CLI

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.jboss.org/browse/WFLY-10832
https://issues.jboss.org/browse/WFLY-7234

Analysis document: https://github.com/wildfly/wildfly-proposals/pull/134

### CODE REVIEW

**Review by kabir — COMMENTED** [review]
@pferraro  I think it looks fine, but I only see tests for 'primary-provider' and not for 'is-primary' and 'providers'

**Review by kabir — APPROVED** [review]

**Comment by kabir** [comment]
@wildfly/mergers-full This looks ok now - there is a question on the EAP7 issue where it is worth checking the reply before merging

**Comment by kabir** [comment]
Retest this please

File: clustering/server/src/main/java/org/wildfly/clustering/server/singleton/DistributedSingletonServiceConfigurator.java [hunk 1]
```diff
@@ -1,5 +1,4 @@
 /*
- * JBoss, Home of Professional Open Source.
```
Comment (jamezp) [inline clustering/server/src/main/java/org/wildfly/clustering/server/singleton/DistributedSingletonServiceConfigurator.java]: Not a big deal probably, but is there a reason this was deleted?

File: clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonServiceResourceDefinition.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        return pathElement(name.getCanonicalName());
+    }
+
+    static PathElement pathElement(String name) {
+        return PathElement.pathElement("service", name);
+    }
+
+    static void buildTransformation(ModelVersion version, ResourceTransformationDescriptionBuilder parent) {
+        if (SingletonModel.VERSION_3_0_0.requiresTransformation(version)) {
+            parent.discardChildResource(WILDCARD_PATH);
```
Comment (jamezp) [inline clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonServiceResourceDefinition.java]: `VERSION_2_0_0` rejects the children, is it safe to discard them here or should they also be rejected?

File: clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonDeploymentResourceDefinition.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        return pathElement(name.getParent().getSimpleName());
+    }
+
+    static PathElement pathElement(String name) {
+        return PathElement.pathElement("deployment", name);
+    }
+
+    static void buildTransformation(ModelVersion version, ResourceTransformationDescriptionBuilder parent) {
+        if (SingletonModel.VERSION_3_0_0.requiresTransformation(version)) {
+            parent.discardChildResource(WILDCARD_PATH);
```
Comment (jamezp) [inline clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonDeploymentResourceDefinition.java]: `VERSION_2_0_0` rejects the children, is it safe to discard them here or should they also be rejected?

**Comment by kabir** [comment]
@pferraro We're getting compilation errors (I think I saw you say somewhere it has been a while since it was rebased)

**Comment by pferraro** [comment]
Thanks for the heads up - will fix today.

**Comment by kabir** [comment]
@pferraro The SingletonPolicyServiceTestCase failure looks a bit suspicious? @jmesnil is having a look at the failures caused by metrics stuff (SecurityAPITestCase etc.)

**Comment by pferraro** [comment]
@kabir Hmm - that's a security-related failure.  I haven't the slightest idea why this fails sometimes.  If there was a genuine problem, it should *always* fail, which is not the case.  Seems like one of the CI instances has an anomalous security configuration - and in that way, it's a matter of which CI worker gets assigned.

File: clustering/server/src/main/java/org/wildfly/clustering/server/singleton/DistributedSingletonServiceConfigurator.java (hunk 1 above)
Comment (pferraro) [reply clustering/server/src/main/java/org/wildfly/clustering/server/singleton/DistributedSingletonServiceConfigurator.java]: I think that may have been the cat walking across my keyboard while I was making coffee. I'll fix.

File: clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonServiceResourceDefinition.java (hunk 2 above)
Comment (pferraro) [reply clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonServiceResourceDefinition.java]: @jamezp Where do you see children being rejected in VERSION_2_0_0?

File: clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonDeploymentResourceDefinition.java (hunk 3 above)
Comment (pferraro) [reply clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonDeploymentResourceDefinition.java]: @jamezp Where do you see children being rejected in VERSION_2_0_0?

**Comment by kabir** [comment]
Retest this please

File: clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonServiceResourceDefinition.java (hunk 2 above)
Comment (jamezp) [reply clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonServiceResourceDefinition.java]: In the [`SingletonPolicyResourceDefinition`](https://github.com/wildfly/wildfly/blob/a2ae8f2762012a2b826641f29dadd28be4420225/clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonPolicyResourceDefinition.java#L123-L127l).

Comment (jamezp) [reply clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonServiceResourceDefinition.java]: Never mind. Kabir pointed out this is a runtime resource so it's probably okay.

File: clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonDeploymentResourceDefinition.java (hunk 3 above)
Comment (jamezp) [reply clustering/singleton/extension/src/main/java/org/wildfly/extension/clustering/singleton/SingletonDeploymentResourceDefinition.java]: Never mind. Kabir pointed out this is a runtime resource so it's probably okay.

**Review by jamezp — APPROVED** [review]

**Pull request merged**